crates/net/network/src/transactions.rs

```diff
@@ -1074,116 +1271,341 @@ struct Peer {
-                    Poll::Ready(FetchEvent::FetchError {
-                        peer_id,
-                        error: RequestError::ChannelClosed,
-                    })
+            }
+        }
+
+        None
+    }
+
+    /// Packages hashes for [`GetPooledTxRequest`] up to limit. Returns left over hashes.
+    fn pack_hashes(&mut self, hashes: &mut Vec<TxHash>, peer_id: PeerId) -> Option<Vec<TxHash>> {
+        let hash = hashes.first()?;
+        if self.eth68_meta.get(hash).is_some() {
+            return self.pack_hashes_eth68(hashes, peer_id)
+        }
+        self.pack_hashes_eth66(hashes, peer_id).map(|left_over| left_over.collect())
+    }
+
+    /// Packages hashes for [`GetPooledTxRequest`] up to limit as defined by protocol version 66.
+    /// If necessary, takes hashes from buffer for which peer is listed as fallback peer.
+    ///
+    /// Returns left over hashes.
+    fn pack_hashes_eth66<'a>(
+        &mut self,
+        hashes: &'a mut Vec<TxHash>,
+        peer_id: PeerId,
+    ) -> Option<impl Iterator<Item = TxHash> + 'a> {
+        if hashes.len() < GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES {
+            self.fill_request_for_peer(hashes, peer_id, None);
+            return None
+        }
+        Some(hashes.drain(..GET_POOLED_TRANSACTION_SOFT_LIMIT_NUM_HASHES))
+    }
+
+    /// Returns `Ok(true)` if hash is included in request. If there is still space in the
+    /// respective response but not enough for the tx of given hash, `Ok(false)` is returned.
+    /// Throws error if [`MAX_FULL_TRANSACTIONS_PACKET_BYTE_SIZE`] is reached.
+    fn include_eth68_hash(
+        &self,
+        acc_tx_size: &mut usize,
+        eth68_hash: TxHash,
+    ) -> Result<bool, &'static str> {
+        debug_assert!(
+            self.eth68_meta.peek(&eth68_hash).is_some(),
+            "broken invariant `eth68-hash` and `eth68-meta`"
+        );
+
+        if let Some(size) = self.eth68_meta.peek(&eth68_hash) {
+            if *size <= MAX_FULL_TRANSACTIONS_PACKET_BYTE_SIZE - *acc_tx_size {
+                // tx will fit into response, include hash in request
+                *acc_tx_size += *size;
+                if *acc_tx_size >= MAX_FULL_TRANSACTIONS_PACKET_BYTE_SIZE {
+                    return Err("limit reached")
                 }
+                return Ok(true)
             }
         }
-        Poll::Pending
+        Ok(false)
+    }
+
+    /// Packages hashes for [`GetPooledTxRequest`] up to limit as defined by protocol version 68.
+    /// If necessary, takes hashes from buffer for which peer is listed as fallback peer.
+    ///
+    /// Returns left over hashes.
+    fn pack_hashes_eth68(
+        &mut self,
+        hashes: &mut Vec<TxHash>,
+        peer_id: PeerId,
+    ) -> Option<Vec<TxHash>> {
+        let mut acc_size = 0;
+        let mut surplus_hashes = vec![];
+
+        for &hash in hashes.iter() {
+            match self.include_eth68_hash(&mut acc_size, hash) {
```

Comment:
perhaps this loop can be solved with a `retain` over `hashes`:

keep all hashes that fit into request, remove the others and push to surplus_hashes

Reply:
oh yeah, at least I should have retained the `surplus_hashes` before filling the request from buffer. can definitely  combine retain and `include_eth68_hash` into one.